### PR TITLE
Replace the metadata collector with tests

### PR DESCRIPTION
**Pull request by Alexendoo**
The metadata collector handles 3 files: [`CHANGELOG.md`](https://github.com/rust-lang/rust-clippy/blob/c082bc2cb85313901ed3565fcd285592ed93df0f/CHANGELOG.md#L6050), [`lint_configuration.md`](https://github.com/rust-lang/rust-clippy/blob/c082bc2cb85313901ed3565fcd285592ed93df0f/book/src/lint_configuration.md) and `util/gh-pages/lints.json`

- `CHANGELOG.md` and `lint_configuration.md` are now checked by `tests/config-metadata.rs`, when they are outdated `cargo test` will fail with a message to run `cargo bless --test config-metadata` in order to update them

   A plain `cargo bless` will run all the tests, blessing both this and the UI tests

- `util/gh-pages/lints.json` is now generated when running `cargo uitest` with `COLLECT_METADATA=1` (still aliased to `cargo collect-metadata`)

   It uses a `ui_test` [post test action](https://docs.rs/ui_test/latest/ui_test/custom_flags/trait.Flag.html#method.post_test_action) to retrieve the applicability from the actual diagnostics emitted during UI tests

   Example change from the current to new JSON:

   ```diff
      {
        "id": "chars_next_cmp",
   -    "id_span": {
   -      "path": "src/methods/mod.rs",
   -      "line": 891
   -    },
   +    "id_location": "clippy_lints/src/methods/mod.rs#891",
        "group": "style",
        "level": "warn",
   -    "docs": "\n### What it does ... ```",
   +    "docs": "### What it does ... ```\n",
        "version": "pre 1.29.0",
   -    "applicability": {
   -      "is_multi_part_suggestion": false,
   -      "applicability": "Unresolved"
   -    }
   +    "applicability": "MachineApplicable"
      },
   ```

   `Hide whitespace` makes the `compile-test.rs` changes much easier to see

r? @flip1995

changelog: none

### CODE REVIEW

File: tests/config-metadata.rs
Comment (flip1995) [inline tests/config-metadata.rs:1]: File name: Should this be `collect-metadata`? :thinking:

Comment (Alexendoo) [reply tests/config-metadata.rs:1]: It's intentionally renamed as it's separate from `cargo collect-metadata`, `cargo collect-metadata` is now only used for generating the website's `lints.json`

**Review by flip1995 — COMMENTED** [review]
Still need to finalize my review, but overall this change makes sense to me. As the "metadata collection monster" is @xFrednet's baby, I also want to give them a chance to take a look.

File: tests/compile-test.rs [hunk 1]
```diff
@@ -347,3 +385,176 @@ fn ui_cargo_toml_metadata() {
...
+            group: lint.category_str(),
+            level: lint.lint.default_level.as_str(),
+            docs,
+            version: lint.version.unwrap(),
+            applicability,
+        }
+    }
+
+    fn new_deprecated(name: &str, reason: &str, version: &'static str) -> Self {
+        // The reason starts with a lowercase letter and end without a period.
```
Comment (Jarcho) [inline tests/compile-test.rs:540]: nit: should be "ends without"

This is probably my typo from an earlier PR.

File: tests/compile-test.rs [hunk 2]
```diff
@@ -347,3 +385,176 @@ fn ui_cargo_toml_metadata() {
...
+                        let applicability = applicabilities
+                            .entry(lint.to_string())
+                            .or_insert(Applicability::Unspecified);
+                        let applicabilities = diag
+                            .children
+                            .iter()
+                            .flat_map(|child| &child.spans)
+                            .filter_map(|span| span.suggestion_applicability.clone())
+                            .chain(iter::once(applicability.clone()));
+                        *applicability = applicabilities.max_by_key(applicability_ord).unwrap();
```
[the remark this answers is not in the document]
Comment (Alexendoo) [reply tests/compile-test.rs:431]: What do you mean by only appear once?

[the remark this answers is not in the document]
Comment (Jarcho) [reply tests/compile-test.rs:431]: A lint can output multiple diagnostics at different applicability levels. It doesn't look like you have anything preventing `*applicability` from being assigned to more than once per lint.

[the remark this answers is not in the document]
Comment (Alexendoo) [reply tests/compile-test.rs:431]: the `iter::once` includes the previous applicability from the map in the comparison, but it's probably clearer to write it out even if I have to repeat `applicability_ord` a few times

[the remark this answers is not in the document]
Comment (Jarcho) [reply tests/compile-test.rs:431]: Ah, I missed that part.

**Review by xFrednet — APPROVED** [review]
Very nice implementation. I really like how you collected all the information! It's a bit sad to see my metadata collection monster go, but this implementation is way better IMO!

Thank you for putting in the time!

@flip1995, I've reviewed all changes. If you don't have the time, you can `r=me` :D

**Comment by flip1995** [comment]
@bors r=xFrednet

Thanks for reviewing this! I didn't get to review it this weekend (at least not for an in-depth review). I like the new way of testing it though. Great overall improvements!

**Comment by xFrednet** [comment]
It looks like the `metadata_collector` test fails for `i686-unknown-linux-gnu`. We can restrict that test to 64bits and linux to see if that fixes it.

**Comment by Alexendoo** [comment]
👀

Going to print a diff first, ideally it wouldn't be platform dependent

**Comment by Alexendoo** [comment]
```diff
-**Default Value:** `["j", "z", "i", "y", "n", "x", "w"]`
+**Default Value:** `["w", "n", "x", "i", "z", "y", "j"]`
```

Can't say that's what I expected, I'll take a better look at this later

**Comment by Alexendoo** [comment]
oops

@bors try

**Comment by Alexendoo** [comment]
We used `FxHashSet` in `Conf` a few times, I'm guessing the order changed because usize hashes differently on 32-bit. I'll leave an actual fix to https://github.com/rust-lang/rust-clippy/pull/13084 but as a temporary measure I replaced them with `Vec`s

**Review by xFrednet — COMMENTED** [review]
LGTM! Could you copy the changes from https://github.com/rust-lang/rust-clippy/pull/13261 to correct the indention and include `compile_fail` for code blocks?

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Everything looks good to me, I only now noticed the branch name ^^

---

### Death of a monster

Roses are red,
Violets are blue,
The collector,
is now anew.

Farewell my friend,
ḿy biggest lint,
who started this journey,
we're all in.

RIP
*2021-05-05
†2024-08-13

**Pull request merged**